crates/swc_ecma_minifier/src/util/base54.rs

```diff
@@ -1,46 +1,198 @@
-use swc_ecma_ast::IdentExt;
-
 pub(crate) static BASE54_DEFAULT_CHARS: &[u8; 64] =
     b"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ$_0123456789";
 
-/// Note: This returns `a` for 0.
-pub(crate) fn incr_base54(init: &mut usize, chars: &[u8; 64]) -> (usize, String) {
+/// givin a number, return a base54 encoded string
+/// `usize -> [a-zA-Z$_][a-zA-Z$_0-9]*`
+pub(crate) fn encode(init: &mut usize, chars: &[u8; 64], skip_reserved: bool) -> String {
+    if skip_reserved {
+        while init.is_reserved()
+            || init.is_reserved_in_strict_bind()
+            || init.is_reserved_in_strict_mode(true)
+        {
+            *init += 1;
+        }
+    }
+
     let mut n = *init;
 
     *init += 1;
 
-    let mut ret = String::new();
     let mut base = 54;
 
-    n += 1;
+    while n >= base {
+        n -= base;
+        base <<= 6;
+    }
 
-    while n > 0 {
-        n -= 1;
+    let mut ret = String::new();
 
-        let c = chars[n % base] as char;
+    base /= 54;
+    let mut c = chars[n / base] as char;
+    ret.push(c);
 
-        if ret.is_empty() && c.is_digit(10) {
-            return incr_base54(init, chars);
-        }
+    while base > 1 {
+        n %= base;
+        base >>= 6;
+        c = chars[n / base] as char;
 
         ret.push(c);
+    }
+
+    ret
+}
+
+#[allow(unused)]
+pub(crate) fn decode(s: &str) -> usize {
+    let mut ret = 0;
+
+    let mut base = 54;
+
+    for _ in 0..(s.len() - 1) {
+        ret += base;
+        base <<= 6;
+    }
+
+    let mut x = 0;
+
+    for c in s.as_bytes().iter() {
+        x <<= 6;
+        x += reverse_base54(c);
+    }
+
+    ret += x;
+
+    ret
+}
+
+fn reverse_base54(c: &u8) -> usize {
```

Comment:
I think accepting `u8` will be better for performance.

Reply:
The `decode` and `reverse_base54` functions are currently not used anywhere except for testing.

Reply:
I see, then it's fine.
Thanks for the explanation